File: frontend/src/components/memory/memoryStylePresets.ts
/**
 * Memory layout presets — must stay in sync with
 * python-ai/app/agent/backend/memory_style_presets.py
 *
 * AI selects `layout` on CreateMemory style JSON; UI resolves CSS via resolveNodePresentation().
 */

import type { CSSProperties } from 'react'

import type { MemoryNodeStyle } from '../../types/memoryNode'
import { normalizeMemoryIconName } from './memoryNodeIcons'

export type MemoryLayoutPreset =
  | 'accordion'
  | 'outline'
  | 'cards'
  | 'timeline'
  | 'hero'
  | 'quote'
  | 'prose'

export type MemoryStyleVariant = 'default' | 'emphasis' | 'muted' | 'quote'

export type MemoryStyleAccent =
  | 'primary'
  | 'emerald'
  | 'amber'
  | 'rose'
  | 'violet'
  | 'sky'

export const MEMORY_LAYOUT_KEYS: MemoryLayoutPreset[] = [
  'accordion',
  'outline',
  'cards',
  'timeline',
  'hero',
  'quote',
  'prose',
]

export interface MemoryLayoutPresetDef {
  layout: MemoryLayoutPreset
  variant?: MemoryStyleVariant
  collapse_default?: boolean
  show_content_inline?: boolean
  aiHint: string
}

export const MEMORY_LAYOUT_PRESETS: Record<MemoryLayoutPreset, MemoryLayoutPresetDef> = {
  accordion: {
    layout: 'accordion',
    variant: 'default',
    collapse_default: false,
    show_content_inline: false,
    aiHint: 'Multi-level worldview / outline sections; collapsible hierarchy.',
  },
  outline: {
    layout: 'outline',
    variant: 'default',
    collapse_default: true,
    show_content_inline: false,
    aiHint: 'Deep nesting (depth≥2); indented outline like a table of contents.',
  },
  cards: {
    layout: 'cards',
    variant: 'emphasis',
    collapse_default: false,
    show_content_inline: true,
    aiHint: 'Sibling items at same level (characters, parallel settings).',
  },
  timeline: {
    layout: 'timeline',
    variant: 'default',
    collapse_default: false,
    show_content_inline: true,
    aiHint: 'Chronological novel outline or history sections.',
  },
  hero: {
    layout: 'hero',
    variant: 'emphasis',
    collapse_default: false,
    show_content_inline: true,
    aiHint: 'Scope root with overview content (node_kind=both).',
  },
  quote: {
    layout: 'quote',
    variant: 'muted',
    collapse_default: false,
    show_content_inline: true,
    aiHint: 'Short leaf excerpt (<400 chars), pull-quote style.',
  },
  prose: {
    layout: 'prose',
    variant: 'default',
    collapse_default: false,
    show_content_inline: true,
    aiHint: 'Default long-form Markdown leaf; omit style or use this.',
  },
}

export const DEFAULT_SCOPE_LAYOUT: Record<string, MemoryLayoutPreset> = {}

/** Fields mirrored in Python SYNC_PRESET_FIELDS / preset_sync_manifest(). */
export const SYNC_PRESET_FIELDS = [
  'layout',
  'variant',
  'collapse_default',
  'show_content_inline',
] as const

export function presetSyncManifest(): Array<Record<(typeof SYNC_PRESET_FIELDS)[number], unknown>> {
  return MEMORY_LAYOUT_KEYS.map((key) => {
    const preset = MEMORY_LAYOUT_PRESETS[key]
    return {
      layout: preset.layout,
      variant: preset.variant,
      collapse_default: preset.collapse_default,
      show_content_inline: preset.show_content_inline,
    }
  })
}

export interface NodePresentation {
  layout: MemoryLayoutPreset
  containerClass: string
  titleClass: string
  contentClass: string
  childContainerClass: string
  showContentInline: boolean
  collapseDefault: boolean
  icon?: string
  accentClass: string
  accentStyle?: CSSProperties
  levelIndentClass: string
}

const LAYOUT_CONTAINER: Record<MemoryLayoutPreset, string> = {
  accordion: 'rounded-xl border border-border/50 bg-card/30 overflow-hidden shadow-sm',
  outline: 'border-l-2 border-primary/25 pl-3',
  cards: 'rounded-xl border border-border/45 bg-card/60 shadow-sm',
  timeline: 'relative pl-4 before:absolute before:left-[5px] before:top-1 before:bottom-1 before:w-px before:bg-border/80',
  hero: 'rounded-xl border border-border/40 bg-gradient-to-br from-muted/50 via-card/40 to-transparent px-4 py-3',
  quote: 'rounded-lg border-l-[3px] border-primary/30 bg-muted/20 px-3 py-2.5',
  prose: 'rounded-xl bg-muted/15 px-3 py-2.5 ring-1 ring-border/40',
}

const LAYOUT_TITLE: Record<MemoryLayoutPreset, string> = {
  accordion: 'text-[14px] font-semibold text-foreground px-3 py-2 border-b border-border/40 bg-muted/20',
  outline: 'text-[14px] font-semibold text-foreground',
  cards: 'text-[14px] font-semibold text-foreground',
  timeline: 'text-[14px] font-semibold text-foreground',
  hero: 'text-[16px] font-bold tracking-tight text-foreground',
  quote: 'text-[13px] font-medium text-foreground not-italic',
  prose: 'text-[14px] font-semibold text-foreground',
}

const ACCENT_BORDER_CLASS: Record<MemoryStyleAccent, string> = {
  primary: 'border-primary/50',
  emerald: 'border-emerald-500/50',
  amber: 'border-amber-500/50',
  rose: 'border-rose-500/50',
  violet: 'border-violet-500/50',
  sky: 'border-sky-500/50',
}

export function normalizeLayout(raw?: string | null): MemoryLayoutPreset {
  const key = (raw || 'prose').trim().toLowerCase() as MemoryLayoutPreset
  return key in MEMORY_LAYOUT_PRESETS ? key : 'prose'
}

export function resolveDefaultLayout(
  style: MemoryNodeStyle | null | undefined,
  opts: { scope?: string; depth?: number; nodeKind?: string; isRoot?: boolean },
): MemoryLayoutPreset {
  if (style?.layout) {
    return normalizeLayout(style.layout)
  }
  if (opts.isRoot && (opts.nodeKind === 'both' || opts.nodeKind === 'section')) {
    return 'hero'
  }
  if ((opts.depth ?? 0) >= 2) {
    return 'outline'
  }
  if (opts.scope && DEFAULT_SCOPE_LAYOUT[opts.scope]) {
    return DEFAULT_SCOPE_LAYOUT[opts.scope]
  }
  return 'prose'
}

function resolveAccent(accent?: string | null): { accentClass: string; accentStyle?: CSSProperties } {
  if (!accent) {
    return { accentClass: '' }
  }
  const trimmed = accent.trim()
  const named = trimmed.toLowerCase() as MemoryStyleAccent
  if (named in ACCENT_BORDER_CLASS) {
    return { accentClass: ACCENT_BORDER_CLASS[named] }
  }
  if (/^#[0-9a-fA-F]{3,8}$/.test(trimmed)) {
    return { accentClass: 'border-opacity-80', accentStyle: { borderColor: trimmed } }
  }
  return { accentClass: '' }
}

function variantContainerClass(variant: MemoryStyleVariant): string {
  if (variant === 'emphasis') {
    return ' ring-1 ring-primary/10'
  }
  if (variant === 'muted') {
    return ' opacity-95'
  }
  return ''
}

function variantTitleClass(variant: MemoryStyleVariant): string {
  if (variant === 'muted') {
    return ' text-muted-foreground'
  }
  return ''
}

function levelIndentClass(level?: number | null): string {
  if (level == null || level <= 0) {
    return ''
  }
  const capped = Math.min(8, Math.max(0, level))
  const map: Record<number, string> = {
    1: 'ml-1',
    2: 'ml-2',
    3: 'ml-3',
    4: 'ml-4',
    5: 'ml-5',
    6: 'ml-6',
    7: 'ml-7',
    8: 'ml-8',
  }
  return map[capped] ?? ''
}

export function resolveNodePresentation(
  style: MemoryNodeStyle | null | undefined,
  opts: { scope?: string; depth?: number; nodeKind?: string; isRoot?: boolean },
): NodePresentation {
  const layout = resolveDefaultLayout(style, opts)
  const preset = MEMORY_LAYOUT_PRESETS[layout]
  const variant = (style?.variant ?? preset.variant ?? 'default') as MemoryStyleVariant
  const { accentClass, accentStyle } = resolveAccent(style?.accent)
  const icon = normalizeMemoryIconName(style?.icon) ?? undefined
  const levelIndent = levelIndentClass(style?.level ?? null)

  return {
    layout,
    containerClass: LAYOUT_CONTAINER[layout] + variantContainerClass(variant) + (accentClass ? ` ${accentClass}` : ''),
    titleClass: LAYOUT_TITLE[layout] + variantTitleClass(variant),
    contentClass:
      'text-[13px] leading-relaxed text-foreground agent-prose agent-prose-memory' +
      (variant === 'muted' ? ' text-muted-foreground' : ''),
    childContainerClass:
      layout === 'cards'
        ? 'grid gap-2 sm:grid-cols-2 mt-2'
        : layout === 'timeline'
          ? 'mt-2 flex flex-col gap-3'
          : 'mt-2 flex flex-col gap-2 ml-0.5',
    showContentInline: style?.show_content_inline ?? preset.show_content_inline ?? true,
    collapseDefault: style?.collapse_default ?? preset.collapse_default ?? false,
    icon,
    accentClass,
    accentStyle,
    levelIndentClass: levelIndent,
  }
}

/** Markdown block for AI tool docs (mirrors Python memory_style_prompt_block). */
export function memoryStyleGuideForDocs(): string {
  return Object.entries(MEMORY_LAYOUT_PRESETS)
    .map(([k, v]) => `- \`${k}\`: ${v.aiHint}`)
    .join('\n')
}

File: python-ai/app/agent/backend/memory_style_presets.py
"""Memory node layout presets — canonical for AI prompts and frontend rendering.

See docs/TOOL_CONTRACT_AND_MEMORY_REFACTOR_PLAN.md §3
"""

from __future__ import annotations

import re
from typing import Any

# layout values must stay in sync with frontend/src/components/memory/memoryStylePresets.ts

MEMORY_LAYOUT_KEYS: tuple[str, ...] = (
    "accordion",
    "outline",
    "cards",
    "timeline",
    "hero",
    "quote",
    "prose",
)

VALID_VARIANTS: frozenset[str] = frozenset({"default", "emphasis", "muted", "quote"})

VALID_ACCENTS: frozenset[str] = frozenset(
    {"primary", "emerald", "amber", "rose", "violet", "sky"}
)

# Lucide PascalCase names — sync with frontend/src/components/memory/memoryNodeIcons.tsx
VALID_MEMORY_ICONS: frozenset[str] = frozenset(
    {
        "BookMarked",
        "BookOpen",
        "Bookmark",
        "Brain",
        "Calendar",
        "Clock",
        "Compass",
        "Contact",
        "Crown",
        "Drama",
        "Eye",
        "Feather",
        "FileText",
        "Flag",
        "Flame",
        "Folder",
        "FolderOpen",
        "Gem",
        "Globe",
        "Heart",
        "History",
        "Landmark",
        "Layers",
        "List",
        "ListTree",
        "Map",
        "MapPin",
        "PenLine",
        "Scroll",
        "ScrollText",
        "Shield",
        "Sparkles",
        "Star",
        "Sword",
        "Target",
        "User",
        "UserCircle",
        "Users",
        "Zap",
    }
)

MEMORY_LAYOUT_PRESETS: dict[str, dict[str, Any]] = {
    "accordion": {
        "layout": "accordion",
        "variant": "default",
        "collapse_default": False,
        "show_content_inline": False,
        "ai_hint": "Multi-level worldview / outline sections; collapsible hierarchy.",
    },
    "outline": {
        "layout": "outline",
        "variant": "default",
        "collapse_default": True,
        "show_content_inline": False,
        "ai_hint": "Deep nesting (depth≥2); indented outline like a table of contents.",
    },
    "cards": {
        "layout": "cards",
        "variant": "emphasis",
        "collapse_default": False,
        "show_content_inline": True,
        "ai_hint": "Sibling items at same level (characters, parallel settings).",
    },
    "timeline": {
        "layout": "timeline",
        "variant": "default",
        "collapse_default": False,
        "show_content_inline": True,
        "ai_hint": "Chronological novel outline or history sections.",
    },
    "hero": {
        "layout": "hero",
        "variant": "emphasis",
        "collapse_default": False,
        "show_content_inline": True,
        "ai_hint": "Scope root with overview content (node_kind=both).",
    },
    "quote": {
        "layout": "quote",
        "variant": "muted",
        "collapse_default": False,
        "show_content_inline": True,
        "ai_hint": "Short leaf excerpt (<400 chars), pull-quote style.",
    },
    "prose": {
        "layout": "prose",
        "variant": "default",
        "collapse_default": False,
        "show_content_inline": True,
        "ai_hint": "Default long-form Markdown leaf; omit style or use this.",
    },
}

DEFAULT_SCOPE_LAYOUT: dict[str, str] = {}

# Fields mirrored in frontend MEMORY_LAYOUT_PRESETS (see test_memory_style_presets.py)
SYNC_PRESET_FIELDS: tuple[str, ...] = (
    "layout",
    "variant",
    "collapse_default",
    "show_content_inline",
)


def preset_sync_manifest() -> list[dict[str, Any]]:
    """Export preset fields for cross-language sync tests."""
    out: list[dict[str, Any]] = []
    for key in MEMORY_LAYOUT_KEYS:
        preset = MEMORY_LAYOUT_PRESETS[key]
        out.append({field: preset.get(field) for field in SYNC_PRESET_FIELDS})
    return out


def _strip_ai_hint(preset: dict[str, Any]) -> dict[str, Any]:
    cleaned = dict(preset)
    cleaned.pop("ai_hint", None)
    return cleaned


def default_style_for_scope(
    scope: str,
    *,
    node_kind: str = "both",
    is_root: bool = False,
    depth: int = 0,
) -> dict[str, Any]:
    """Suggest style when AI omits layout on CreateMemory / UpdateMemoryFields."""
    if is_root and node_kind in ("both", "section"):
        return _strip_ai_hint(MEMORY_LAYOUT_PRESETS["hero"])
    if depth >= 2:
        return _strip_ai_hint(MEMORY_LAYOUT_PRESETS["outline"])
    s = (scope or "").strip().lower()
    key = DEFAULT_SCOPE_LAYOUT.get(s, "prose")
    return _strip_ai_hint(MEMORY_LAYOUT_PRESETS.get(key, MEMORY_LAYOUT_PRESETS["prose"]))


def _coerce_variant(raw: Any) -> str | None:
    if raw is None:
        return None
    variant = str(raw).strip().lower()
    return variant if variant in VALID_VARIANTS else "default"


def _coerce_accent(raw: Any) -> str | None:
    if raw is None:
        return None
    accent = str(raw).strip()
    if not accent:
        return None
    if accent.startswith("#") and 4 <= len(accent) <= 9:
        return accent
    lowered = accent.lower()
    return lowered if lowered in VALID_ACCENTS else None


def _to_lucide_icon_key(name: str) -> str:
    text = name.strip()
    if not text:
        return ""
    if text.isascii() and text[0].isupper() and "-" not in text and "_" not in text and " " not in text:
        return text
    parts = [p for p in re.split(r"[-_\s]+", text) if p]
    return "".join(p[:1].upper() + p[1:] for p in parts if p)


def _coerce_icon(raw: Any) -> str | None:
    if raw is None:
        return None
    text = str(raw).strip()
    if not text or not text.isascii():
        return None
    key = _to_lucide_icon_key(text)
    if key in VALID_MEMORY_ICONS:
        return key
    return None


def normalize_style(
    raw: dict[str, Any] | None,
    *,
    scope: str = "",
    node_kind: str = "leaf",
    depth: int = 0,
    is_root: bool = False,
) -> dict[str, Any]:
    """Merge AI/user style JSON with preset defaults; unknown layout → prose."""
    if not isinstance(raw, dict) or not raw.get("layout"):
        base = default_style_for_scope(scope, node_kind=node_kind, is_root=is_root, depth=depth)
        if isinstance(raw, dict):
            icon = _coerce_icon(raw.get("icon"))
            if icon:
                base["icon"] = icon
            elif "icon" in raw:
                base.pop("icon", None)
        return base
    layout = str(raw.get("layout") or "prose").strip().lower()
    if layout not in MEMORY_LAYOUT_PRESETS:
        layout = "prose"
    base = _strip_ai_hint(MEMORY_LAYOUT_PRESETS[layout])
    for key, value in raw.items():
        if key == "ai_hint" or value is None:
            continue
        base[key] = value
    base["layout"] = layout
    variant = _coerce_variant(base.get("variant"))
    if variant is not None:
        base["variant"] = variant
    accent = _coerce_accent(base.get("accent"))
    if "accent" in raw:
        base["accent"] = accent
    if "icon" in raw or "icon" in base:
        icon = _coerce_icon(base.get("icon"))
        if icon:
            base["icon"] = icon
        else:
            base.pop("icon", None)
    level = base.get("level")
    if level is not None:
        try:
            base["level"] = max(0, min(8, int(level)))
        except (TypeError, ValueError):
            base.pop("level", None)
    return base


def memory_style_prompt_block() -> str:
    lines = [
        "## Memory node style presets (set on CreateMemory / UpdateMemoryFields `style` JSON)",
        "Pick `layout` from the table. Optional fields:",
        "- `variant`: default | emphasis | muted | quote",
        "- `icon`: Lucide icon name (PascalCase), rendered by frontend — no emoji. "
        f"Allowed: {', '.join(sorted(VALID_MEMORY_ICONS))}",
        "- `accent`: primary | emerald | amber | rose | violet | sky | #RRGGBB (border tint)",
        "- `collapse_default`, `show_content_inline`: bool overrides",
        "",
        "| layout | When to use |",
        "|--------|-------------|",
    ]
    for key in MEMORY_LAYOUT_KEYS:
        preset = MEMORY_LAYOUT_PRESETS[key]
        hint = str(preset.get("ai_hint") or "")
        lines.append(f"| `{key}` | {hint} |")
    lines.extend(
        [
            "",
            "Rules:",
            "- Two-level UI: scope root = left tab; children = sub-menu; one child body per panel",
            "- Put readable Markdown on **child** nodes; scope root = title + optional short intro only",
            "- Split large topics into multiple CreateMemory(node_type=child) instead of one giant root/update",
            "- section nodes: accordion or outline (depth≥2 nested sections)",
            "- leaf long Markdown: omit style or layout=prose",
            "- sibling parallel items: layout=cards, variant=emphasis",
            "- scope root (CreateMemory node_type=root): layout=hero + node_kind=both; title becomes scope tab",
            "- short leaf (<400 chars): layout=quote, variant=muted",
            "- style.icon must be a Lucide name from the allowlist (never emoji)",
            "- unknown layout values are stored as prose",
        ]
    )
    return "\n".join(lines)
